valohai_cli/utils/ssh.py

```diff
@@ -0,0 +1,147 @@
+import json
+import os
+import pathlib
+import shutil
+import subprocess
+from dataclasses import dataclass
+from typing import Generator, Iterable, Iterator, Optional, Tuple
+
+import click
+
+from valohai_cli.api import request
+from valohai_cli.ctx import get_project
+from valohai_cli.exceptions import ExecutionNotRunning
+from valohai_cli.settings import get_settings_file_name
+from valohai_cli.utils import call_with_retry
+from valohai_cli.utils.cli_utils import prompt_from_list
+
+
+@dataclass(frozen=True)
+class Keypair:
+    public_path: pathlib.Path
+    private_path: pathlib.Path
+
+
+def make_ssh_connection(address: str, port: int, private_key_path: str) -> None:
+    ssh_path = shutil.which("ssh")
+    if ssh_path is None:
+        raise FileNotFoundError(
+            "SSH command not found. Ensure SSH is installed and available in your PATH.",
+        )
+    ssh_command = [
+        ssh_path,
+        "-i",
+        private_key_path,
+        f"root@{address}",
+        "-p",
+        str(port),
+        "-t",
+        "/bin/bash",
+    ]
+    subprocess.run(ssh_command)
+
+
+def find_ssh_details(events: Iterable) -> Optional[Tuple[str, int]]:
+    prefix = "::ssh::"
+    for event in events:
+        if event["message"].startswith(prefix):
+            data = json.loads(event["message"][len(prefix) :])
+            return data["address"], int(data["port"])
+    return None
+
+
+def get_ssh_key_pairs_from_dir(dir: str) -> Iterator[Keypair]:
+    folder_path = pathlib.Path(dir)
+    for pth in folder_path.glob("*.pub"):
+        if not pth.is_file():
+            continue
+        try:
+            with pth.open("rb") as f:
+                first_line = f.readline().strip()
+                if first_line.startswith((b"ssh-rsa", b"ecdsa-sha2-", b"ssh-ed25519")):
+                    private_key_paths = [pth.with_suffix(""), pth.with_suffix(".pem")]
+                    for private_key_path in private_key_paths:
+                        if private_key_path.is_file():
+                            yield Keypair(public_path=pth, private_path=private_key_path)
+                            break
+        except OSError:  # e.g. unreadable key
+            pass
+
+
+def select_private_key_from_possible_directories() -> Optional[str]:
```

Comment:
Isn't the rtype always `str`? This either raises or returns a string.
```suggestion
def select_private_key_from_possible_directories() -> str:
```